### ISSUE TITLE
Training requires too much RAM

### ISSUE THREAD

**Issue by stweil** [Calamari-OCR/calamari#285]
Training with a large number of lines requires a huge amount of RAM (52 GiB RAM for 375000 lines). Loading the samples into memory contributes only a smaller part to this. Most memory is used by the following data processors (CenterNormalizerProcessor, FinalPreparation, BidiTextProcessor, ...). It looks like the memory which is used for these processors is not released in later steps. Maybe that can be changed by throwing away data which is no longer used.

Reducing the memory requirements is especially important for `calamari-cross-fold-train` which currently has to be restricted to a subset of folds even on a large server with 128 GiB of RAM.

**Comment by ChWick** [comment]
Do you use data augmentation?

There is a `preload` flag that can disable preloading the files into RAM:
```
--train.preload=False
--val.preload=False
```

**Comment by stweil** [comment]
No, I don't use data augmentation at the moment. I saw the `preload` flag, but think that normally preloading is good as otherwise the data would have to be loaded again and again for each epoch. Am I wrong?

I think the central point is not preloading, but keeping intermediate data from the data processors. But of course I still don't now the internals of Calamari, so I might be wrong with that assumption.

**Comment by mikegerber** [comment]
With Calamari 2.3.1 and training on ~ 480k lines (cross fold, max parallel models 1), using TensorFlow 2.14, I observe roughly 174GB(!) memory usage. Here is the output of smem, and the number is the sum of the PSS column here:

```
$ smem -k -P calamari
  PID User     Command                         Swap      USS      PSS      RSS
136749 mike.gerber /bin/bash ./train-calamari2        0   288.0K   301.0K     2.3M
136757 mike.gerber tee model-dta-gt-small-cala        0   328.0K   340.0K     2.1M
187592 mike.gerber /usr/bin/python3 /usr/bin/s        0    15.6M    16.7M    21.7M
175573 mike.gerber /data-ssd/mike.gerber/micro        0    76.4G    76.4G    76.5G
136756 mike.gerber /data-ssd/mike.gerber/micro     8.0K    98.1G    98.1G    98.2G
```

Using TensorFlow 2.15, I see "only" 77 GB of memory usage:

```
$ smem -k -P calamari
  PID User     Command                         Swap      USS      PSS      RSS
216062 mike.gerber sh train-calamari2.sh              0   116.0K   160.0K     1.7M
216070 mike.gerber tee model-dta-gt-small-cala        0   344.0K   357.0K     2.1M
220831 mike.gerber /usr/bin/python3 /usr/bin/s        0    15.6M    16.7M    21.7M
216069 mike.gerber /data-ssd/mike.gerber/micro        0   336.1M   427.0M   520.7M
216140 mike.gerber /data-ssd/mike.gerber/micro        0    76.1G    76.2G    76.3G
```

So, in my case, this is/seems to be at least partly a memory leak in TensorFlow. 

I tried later versions of TensorFlow, but couldn't get Calamari to work on those quickly.

**Comment by andbue** [comment]
Thanks for narrowing that down! We tried to do some memory profiling a few months ago but I still could not find out where all the memory is going. My last guess was that under some conditions some versions of opencv might leak in combination with queue.Queue, but maybe not with queue.SimpleQueue so I built that into branch oom_fix. I don't know if it works, in the end we didn't have the time to test it. If you've got the setup already it might be worth a try to see if https://github.com/Calamari-OCR/calamari/commit/a79fe10653a40cf0590056775969b41ec13762d2 makes a difference for you. Other than that it seems that upgrading tfaip to use a more recent version of tf might be the easiest solution.

**Comment by mikegerber** [comment]
> Thanks for narrowing that down! We tried to do some memory profiling a few months ago but I still could not find out where all the memory is going. My last guess was that under some conditions some versions of opencv might leak in combination with queue.Queue, but maybe not with queue.SimpleQueue so I built that into branch oom_fix. I don't know if it works, in the end we didn't have the time to test it. If you've got the setup already it might be worth a try to see if [a79fe10](https://github.com/Calamari-OCR/calamari/commit/a79fe10653a40cf0590056775969b41ec13762d2) makes a difference for you. Other than that it seems that upgrading tfaip to use a more recent version of tf might be the easiest solution.

I'll test it - While training is done for now it's easy to run it again and the memory use should be visible within a few minutes.

@andbue Should Calamari work with TensorFlow >= 2.16? I'm asking because I'd like to test that first TBH.

**Comment by andbue** [comment]
I think that @bertsky has digged the deepest into tfaip. If I remember correctly, it would need some changes to get it running with numpy >=1.2, then there is an open PR for adabelief (https://github.com/juntang-zhuang/Adabelief-Optimizer/pull/68), and API changes for optimizers and other stuff in TF/keras?

**Comment by mikegerber** [comment]
> I think that [@bertsky](https://github.com/bertsky) has digged the deepest into tfaip. If I remember correctly, it would need some changes to get it running with numpy >=1.2, then there is an open PR for adabelief ([juntang-zhuang/Adabelief-Optimizer#68](https://github.com/juntang-zhuang/Adabelief-Optimizer/pull/68)), and API changes for optimizers and other stuff in TF/keras?

I see, I guess I'll stick to the earlier TF version for now and see what I can do to make this work for us.

**Comment by mikegerber** [comment]
Just to rule out the other possible TF versions: With TF 2.12 and 2.13 I also get 70-80GB of memory usage. Didn't check TF 2.11 because that would have required downgrading Python (at least with my Conda/micromamba setup).

Next up is checking @andbue's oom_fix branch.

| TF | memory use
| - |-
|2.12-2.13 | up to 80 GB
|2.14 | up to 175 GB
|2.15 | up to 80 GB

I feel extremly lucky having used the clear winner here (2.14) first 🥇
